### PR TITLE
fix(curriculum): add an explicit no funds test

### LINKED ISSUE #56241: Cash register: Test 17 fails sometimes
### Describe the Issue

Test 17 is incorrect sometimes, it fails around 10% of the time.

One incorrect case is
```
cash = 50
price = 22.7
cid = [
  [ 'PENNY', 0 ],
  [ 'NICKEL', 0 ],
  [ 'DIME', 0.7 ],
  [ 'QUARTER', 2 ],
  [ 'ONE', 1 ],
  [ 'FIVE', 55 ],
  [ 'TEN', 70 ],
  [ 'TWENTY', 140 ],
  [ 'ONE HUNDRED', 0 ]
]
```
result is `OPEN TWENTY: $20.00 FIVE: $5.00 ONE: $1.00 QUARTER: $1.00 DIME: $0.30`
which is correct based on the input, but it's supposed to be `INSUFFICIENT_FUNDS` according to the test

### Affected Page

https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures-v8/build-a-cash-register-project/build-a-cash-register

### Your code

```js

let price = 79.35;
let cid = [ [ 'PENNY', 0.03 ],
  [ 'NICKEL', 0 ],
  [ 'DIME', 0.9 ],
  [ 'QUARTER', 3.75 ],
  [ 'ONE', 7 ],
  [ 'FIVE', 70 ],
  [ 'TEN', 90 ],
  [ 'TWENTY', 200 ],
  [ 'ONE HUNDRED', 1400 ] ];

const CurrencyValue = [
  1,
  5,
  10,
  25,
  100,
  500,
  1000,
  2000,
  10000,
];

const cashInput = document.getElementById("cash");
const purchaseButton = document.getElementById("purchase-btn");
const changeDueSpan = document.getElementById("change-due");

function calcChangeDue(cash) {
  function currenciesToString(arr) {
    return arr.reduce((str, val) => str + (val[1] !== 0 ? ` ${val[0]}: $${(val[1] / 100).toFixed(2)}` : ""), "")
  }

  console.log(cid)

  // Uses integers for arithmetic to avoid floating-point errors
  const cash_pennies = Math.round(cash * 100);

  const cid_pennies = cid.map(elem => [elem[0], Math.round(elem[1] * 100)]);
  const totalDrawer = [cid_pennies[0][1]];

  for (let i = 1; i < cid_pennies.length; i++) {
    totalDrawer.push(totalDrawer[i - 1] + cid_pennies[i][1]);
  }

  // Close if drawer contains the exact amount of change requested
  if (totalDrawer[cid_pennies.length - 1] === cash_pennies) {
    return "CLOSED" + currenciesToString(cid_pennies);
  }

  const recur = (idx, cashLeft) => {
    if (totalDrawer[idx] < cashLeft) {
      return null;
    }

    for (let currency = cid_pennies[idx][1]; currency >= 0; currency -= CurrencyValue[idx]) {
      if (currency > cashLeft) continue;

      if (currency == cashLeft) {
        return [[cid_pennies[idx][0], currency]]
      }
      if (idx > 0) {
        if (currency + totalDrawer[idx - 1] < cashLeft) {
          break;
        }

        const recursive = recur(idx - 1, (cashLeft - currency));
        if (recursive != null) {
          return [[cid_pennies[idx][0], currency], ...recursive];
        }
      }
    }

    return null;
  }
  const result = recur(cid.length-1, cash_pennies);

  if (result == null) {
    return "INSUFFICIENT_FUNDS";
  }

  return "OPEN" + currenciesToString(result);
}

purchaseButton.addEventListener("click", () => {
  const cash = Number.parseFloat(cashInput.value);
  console.log(cash, price);
  if (cash < price) {
    alert("Customer does not have enough money to purchase the item");
  } else if (cash === price) {
    changeDueSpan.innerText = "No change due - customer paid with exact cash";
  } else {
    const changeDue = calcChangeDue(cash - price);

    console.log(changeDue);
    
    changeDueSpan.innerText = "Status: " + changeDue;
  }
  console.log();
});

```

### Expected behavior

The individual denomination amounts should make it impossible to return the needed change.

### System

- Device: Desktop
- OS: Windows 10
- Browser: Firefox

### PR DESCRIPTION
**Pull request by a2937**
Checklist:

- [X] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [X] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [X] My pull request targets the `main` branch of freeCodeCamp.
- [X] I have tested these changes either locally on my machine, or GitPod.

Closes #56241

I added the explicit test to both the lab and the current cash register project.

### CODE REVIEW

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/build-a-cash-register.md [hunk 1]
```diff
@@ -48,7 +48,7 @@ Your application should show different messages depending on the price of the it
...
 1. When `price` is `19.5`, the value in the `#cash` element is `20`, `cid` is `[["PENNY", 0.01], ["NICKEL", 0], ["DIME", 0], ["QUARTER", 0], ["ONE", 1], ["FIVE", 0], ["TEN", 0], ["TWENTY", 0], ["ONE HUNDRED", 0]]`, and the `#purchase-btn` element is clicked, the value in the `#change-due` element should be `"Status: INSUFFICIENT_FUNDS"`
 1. When `price` is `19.5`, the value in the `#cash` element is `20`, `cid` is `[["PENNY", 0.5], ["NICKEL", 0], ["DIME", 0], ["QUARTER", 0], ["ONE", 0], ["FIVE", 0], ["TEN", 0], ["TWENTY", 0], ["ONE HUNDRED", 0]]`, and the `#purchase-btn` element is clicked, the value in the `#change-due` element should be `"Status: CLOSED PENNY: $0.5"`
-
+1. When `price` is `22.7`, the value in the `#cash` element is `50`,  `cid` is `[["PENNY", 0], ["NICKEL", 0], ["DIME", 0.7], ["QUARTER", 2], ["ONE", 1], ["FIVE", 55], ["TEN", 70], ["TWENTY", 140], ["ONE HUNDRED", 0]]` ,and the `#purchase-btn` element is clicked, the value in the `#change-due` element should be `"Status: INSUFFICIENT_FUNDS"`
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/build-a-cash-register.md:51]:
```suggestion
1. When `price` is `22.7`, the value in the `#cash` element is `50`, `cid` is `[["PENNY", 0], ["NICKEL", 0], ["DIME", 0.7], ["QUARTER", 2], ["ONE", 1], ["FIVE", 55], ["TEN", 70], ["TWENTY", 140], ["ONE HUNDRED", 0]]`, and the `#purchase-btn` element is clicked, the value in the `#change-due` element should be `"Status: INSUFFICIENT_FUNDS"`
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/build-a-cash-register.md [hunk 2]
```diff
@@ -48,7 +48,7 @@ Your application should show different messages depending on the price of the it
...
 Fulfill the user stories and pass all the tests below to complete this project. Give it your own personal style. Happy Coding!
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/build-a-cash-register.md:52]:
```suggestion

Fulfill the user stories and pass all the tests below to complete this project. Give it your own personal style. Happy Coding!
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/build-a-cash-register.md [hunk 3]
```diff
@@ -289,6 +289,34 @@ const notExpected = [/NICKEL/];
 assert.isTrue(!notExpected.some(regex => result.match(new RegExp(regex, 'i'))));
 ```
 
+When `price` is `22.7`, the value in the `#cash` element is `50`,  `cid` is `[["PENNY", 0], ["NICKEL", 0], ["DIME", 0.7], ["QUARTER", 2], ["ONE", 1], ["FIVE", 55], ["TEN", 70], ["TWENTY", 140], ["ONE HUNDRED", 0]]` ,and the `#purchase-btn` element is clicked, the value in the `#change-due` element should be `"Status: INSUFFICIENT_FUNDS"`
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/build-a-cash-register.md:292]:
```suggestion
When `price` is `22.7`, the value in the `#cash` element is `50`, `cid` is `[["PENNY", 0], ["NICKEL", 0], ["DIME", 0.7], ["QUARTER", 2], ["ONE", 1], ["FIVE", 55], ["TEN", 70], ["TWENTY", 140], ["ONE HUNDRED", 0]]`, and the `#purchase-btn` element is clicked, the value in the `#change-due` element should be `"Status: INSUFFICIENT_FUNDS"`
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/build-a-cash-register.md [hunk 4]
```diff
@@ -289,6 +289,34 @@ const notExpected = [/NICKEL/];
...
+]
+cashInput.dispatchEvent(new Event('change'));
+purchaseBtn.click();
+assert.strictEqual(
+  changeDueDiv.innerText.trim().toLowerCase(),
+  'status: insufficient_funds'
+);
+```
+
+
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/build-a-cash-register.md:319]:
```suggestion

```

File: curriculum/challenges/english/25-front-end-development/lab-cash-register/66c58d2477d7622ddc30ae06.md
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/lab-cash-register/66c58d2477d7622ddc30ae06.md:1]: I thought it was the same project (html+js), as they have different ids they don't need to be kept on pair. Thank you for adapting the test to the different format

**Comment by ilenia-magoni** [comment]
the id of the lab cash register was wrong, you can resolve the merge conflict (keep the new id) or drop the lab from this PR

**Comment by a2937** [comment]
I'll drop the lab

**Comment by DanTGL** [comment]
This doesn't actually fix the issue. The issue is that INSUFFICIENT_FUNDS is wrong.

**Pull request closed without merging**